hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/PoolAlignmentContext.java

```diff
@@ -64,7 +64,11 @@ public void updateResponseState(RpcHeaderProtos.RpcResponseHeaderProto.Builder h
    */
   @Override
   public void receiveResponseState(RpcHeaderProtos.RpcResponseHeaderProto header) {
-    sharedGlobalStateId.accumulate(header.getStateId());
+    if (header.getStateId() == 0 && sharedGlobalStateId.get() > 0) {
```

Comment:
stateId 0 means no state id right? 
This is different than msync being 0?

Reply:
I assume stateId is an integer and protobuf will return 0 for an integer if it is not set?

Reply:
Yes, stateID `0` means no stateId is set. Msync doesn't have a return value.

There is another bug in the router in that it accepts 0 as a value to advance it's cachedStateID to.
Ideally `sharedGlobalStateId.get() > 0` should not be necessary here. For now it captures namenodes that actually had STATE_ID_CONTEXT enabled to begin with. But stale reads could happen with a namenode that has never had STATE_ID_CONTEXT enabled.

Fixing this will touch other tests so I'm debating whether to try fix that in this PR or separately. I'm leaning towards a separate PR.

Reply:
We can remove the `sharedGlobalStateId.get() > 0`, right? 

If sharedGlobalStateId.get()  < 0, routers already fallback to active and no need to reset. If it is > 0 and then we see a request without StateID, we will reset this counter and routers will fallback to active.

Adding `sharedGlobalStateId.get() > 0` doesn't seem to make a difference.

Reply:
The tests in TestNoNamenodesAvailableLongTime rely on the router allowing a stateId of 0. So having `sharedGlobalStateId.get() > 0` allows this behavior which guarding against when the sharedGlobalStateId has advances beyond zero.

The tests in `TestNoNamenodesAvailableLongTime` do need to be fixed but I would like to limit the scope of this PR.